fairscale/optim/oss.py

```diff
@@ -201,6 +202,72 @@ def step(self, closure: Optional[Callable[[], float]] = None, **kwargs: Any) ->
 
         return loss
 
+    def clip_grad_norm(self, max_norm: float, norm_type: float = 2.0) -> torch.Tensor:
+        """
+        Clip all gradients at this point in time. The norm is computed over all gradients together, as if they were
+        concatenated into a single vector. Gradients are modified in-place.
+
+        Arguments:
+            max_norm (float or int): max norm of the gradients
+            norm_type (float or int): type of the used p-norm. Can be ``'inf'`` for
+                infinity norm.
+
+        Returns:
+            Total norm of the parameters (viewed as a single vector).
+
+        .. note: This is analogous to `torch.nn.utils.clip_grad_norm_` but handles the partitioning and multiple devices per rank
+        under the hood.
+
+        .. warning: This needs to be called on all ranks, since synchronization primitives will be used
+        """
+
+        # Compute the max norm for this shards's worth of gradients
+        max_norm = float(max_norm)
+        norm_type = float(norm_type)
+
+        # Filter out the grad-less params, concatenate params from all devices
+        local_params = [
+            filter(lambda x: x.grad is not None, device_params[self.rank])
+            for device_params in self.per_device_params.values()
+        ]
+
+        # Compute the norm on this grad set
+        if norm_type == inf:
+            local_norm = max(
+                p.grad.detach().abs().max().to(self._device)  # type: ignore   # mypy seems wrong here, None grads have been filtered out
+                for local_params_device in local_params
+                for p in local_params_device
+            )
+
+        else:
+            local_norm = torch.norm(  # type: ignore
+                torch.stack(
+                    [
+                        torch.norm(p.grad.detach(), norm_type).to(self._device)  # type: ignore
+                        for local_params_device in local_params
+                        for p in local_params_device
+                    ]
+                ),
+                norm_type,
+            )
+
+        # Sync all the norms from all ranks
+        norms = [torch.zeros_like(local_norm) for k in range(self.world_size)]
+        dist.all_gather(norms, local_norm, group=self.group, async_op=False)
```

Comment:
This gather assumes that each rank in the group has the same gradients. Otherwise, you'll end up with different clipping even if the ranks are the same. This will lead to strange behavior in the model parallelism case that is difficult to track down.

Continuing on comment from above, if we allreduce the model parallelism gradients above, then gradients should be the same for the same rank and your assumption will hold in model parallelism.

Reply:
> This gather assumes that each rank in the group has the same gradients. 

hmm, that was not the intent, each rank computes the norm on the gradient it has (they differ), then they all_gather so that everyone gets the overall max ?